### PR TITLE
add alloy_chains

### LINKED ISSUE #5429: Use `alloy_chains::Chain` instead of vendoring own implementation
### Describe the feature

Use [`alloy_chains::Chain`](https://github.com/alloy-rs/chains) instead of vendoring our own implementation in `reth_primitives::Chain`. Both `Chain` and `NamedChain` should be able to be replaced easily.

### Additional context

`alloy_chains`
- [on crates.io](https://crates.io/crates/alloy-chains)
- [on docs.rs](https://docs.rs/alloy-chains)

### PR DESCRIPTION
**Pull request by 1010adigupta**
resolves #5429

### CODE REVIEW

**Comment by 1010adigupta** [comment]
Hey @mattsse I have just started the work of switching to alloy_chains, just adde few lines now.
Actually I was getting the errors related to some traits which are used for Chain and NamedChain in reth earlier are not implemented for alloy_chains:
Screenshot 2024-01-05 at 5 10 46 PM
so should simultaneously make pr for this on alloy_chains repo as well?? implementing what is necessary there?
or is there any other solution?

File: crates/net/eth-wire/src/errors/eth.rs [hunk 1]
```diff
@@ -101,7 +102,7 @@ pub enum EthHandshakeError {
     MismatchedProtocolVersion(GotExpected<u8>),
     #[error("mismatched chain in status message: {0}")]
     /// Mismatch in chain details in status messages.
-    MismatchedChain(GotExpected<Chain>),
+    MismatchedChain(GotExpected<ChainKind>),
```
Comment (mattsse) [inline crates/net/eth-wire/src/errors/eth.rs:105]: this should remain Chain

File: crates/net/eth-wire/src/types/status.rs [hunk 2]
```diff
@@ -24,7 +25,7 @@ pub struct Status {
 
     /// The chain id, as introduced in
     /// [EIP155](https://eips.ethereum.org/EIPS/eip-155#list-of-chain-ids).
-    pub chain: Chain,
+    pub chain: ChainKind,
```
Comment (mattsse) [inline crates/net/eth-wire/src/types/status.rs:28]: should remain Chain

File: Cargo.toml [hunk 3]
```diff
@@ -150,6 +150,7 @@ alloy-primitives = "0.5"
 alloy-dyn-abi = "0.5"
 alloy-sol-types = "0.5"
 alloy-rlp = "0.3"
+alloy-chains = "0.1.7"
```
Comment (mattsse) [inline Cargo.toml:153]: this needs features:

https://github.com/alloy-rs/chains/blob/a36d918234a40e62f932507448c8c3e1e040be16/src/chain.rs#L110-L175

**Review by mattsse — CHANGES_REQUESTED** [review]
the missing impls are hidden behind features

File: Cargo.toml (hunk 3 above)
Comment (1010adigupta) [reply Cargo.toml:153]: okay got it

File: bin/reth/src/args/network_args.rs [hunk 4]
```diff
@@ -88,6 +88,7 @@ impl NetworkArgs {
         secret_key: SecretKey,
         default_peers_file: PathBuf,
     ) -> NetworkConfigBuilder {
+        //bootnode function is not implemented yet for alloy_chains::Chain
         let chain_bootnodes = chain_spec.chain.bootnodes().unwrap_or_else(mainnet_nodes);
```
Comment (1010adigupta) [inline bin/reth/src/args/network_args.rs:92]: ig the bootnode function is not yet implemented for alloy_chains::Chain, should I make a pr for this on alloy_chains repo, or am I missing something here?

File: bin/reth/src/dirs.rs [hunk 5]
```diff
@@ -12,8 +12,8 @@ use std::{
 /// Constructs a string to be used as a path for configuration and db paths.
 pub fn config_path_prefix(chain: Chain) -> String {
     match chain {
```
Comment (1010adigupta) [inline bin/reth/src/dirs.rs:14]: here whether I use Named of from_named, I am getting the same errror, need some help here?

Comment (supernovahs) [reply bin/reth/src/dirs.rs:14]: I think you can use 
```
match chain {
        name => name.to_string(),
        id => id.to_string(),
    }

```

Comment (supernovahs) [reply bin/reth/src/dirs.rs:14]: I think this can be used

```
pub fn config_path_prefix(chain: Chain) -> String {
    if let Some(named) = chain.named() {
        named.to_string()
    } else {
        chain.id().to_string()
    }
}

```

File: bin/reth/src/args/network_args.rs (hunk 4 above)
Comment (mattsse) [reply bin/reth/src/args/network_args.rs:92]: I don't think we should move this to chains directly.
so instead we should add a ChainSpec::bootnodes function and do a match in there with a helper function

File: bin/reth/src/dirs.rs (hunk 5 above)
Comment (gakonst) [reply bin/reth/src/dirs.rs]: should we move this to a `Display` / `ToString` impl?

Comment (Rjected) [reply bin/reth/src/dirs.rs]: seems like `Chain` already implements `Display` (which gets us `ToString`)

File: crates/net/eth-wire/src/types/status.rs [hunk 6]
```diff
@@ -323,7 +325,7 @@ mod tests {
```
Comment (Rjected) [inline crates/net/eth-wire/src/types/status.rs:325]: The original chainspec in this test should be using `.chain(Chain::from_id(1337))`

File: crates/trie/src/proof.rs [hunk 7]
```diff
@@ -168,7 +168,8 @@ mod tests {
     use once_cell::sync::Lazy;
     use reth_db::database::Database;
     use reth_interfaces::RethResult;
-    use reth_primitives::{Account, Bytes, Chain, ChainSpec, StorageEntry, HOLESKY, MAINNET, U256};
+    use reth_primitives::{Account, Bytes, ChainSpec, StorageEntry, HOLESKY, MAINNET, U256};
+    use alloy_chains::Chain;
     use reth_provider::{test_utils::create_test_provider_factory, HashingWriter, ProviderFactory};
     use std::{str::FromStr, sync::Arc};
```
Comment (Rjected) [inline crates/trie/src/proof.rs:175]: `Chain::Id` usages in this file should be converted to `Chain::from_id`

File: bin/reth/src/init.rs [hunk 8]
```diff
@@ -229,9 +229,10 @@ mod tests {
...
     };
     use reth_primitives::{
-        Address, Chain, ForkTimestamps, Genesis, GenesisAccount, IntegerList, GOERLI,
+        Address, ForkTimestamps, Genesis, GenesisAccount, IntegerList, GOERLI,
         GOERLI_GENESIS_HASH, MAINNET, MAINNET_GENESIS_HASH, SEPOLIA, SEPOLIA_GENESIS_HASH,
     };
+    use alloy_chains::Chain;
     use std::collections::HashMap;
 
     fn collect_table_entries<DB, T>(
```
Comment (Rjected) [inline bin/reth/src/init.rs:238]: `Chain::Id` usages in this file should be converted to `Chain::from_id`

**Review by Rjected — DISMISSED** [review]
needs rebase and some changes to pass tests and satisfy clippy.

for `parse` test errors this should work: `.parse::<alloy_chains::Chain>().unwrap()`

File: bin/reth/src/dirs.rs [hunk 9]
```diff
@@ -11,9 +11,10 @@ use std::{
 
 /// Constructs a string to be used as a path for configuration and db paths.
 pub fn config_path_prefix(chain: Chain) -> String {
-    match chain {
-        Chain::from_named(name) => name.to_string(),
-        Chain::from_id(id) => id.to_string(),
+    if let Some(named) = chain.named() {
+        named.to_string()
+    } else {
+        chain.id().to_string()
```
Comment (mattsse) [inline bin/reth/src/dirs.rs:17]: this is the same as the Display impl, so this can simply call chain.to_string()

**Review by mattsse — CHANGES_REQUESTED** [review]
great, one more nit

File: bin/reth/src/args/network_args.rs [hunk 10]
```diff
@@ -88,7 +88,8 @@ impl NetworkArgs {
         secret_key: SecretKey,
         default_peers_file: PathBuf,
     ) -> NetworkConfigBuilder {
-        let chain_bootnodes = chain_spec.chain.bootnodes().unwrap_or_else(mainnet_nodes);
+        //bootnode function is not implemented yet for alloy_chains::Chain
```
Comment (mattsse) [inline bin/reth/src/args/network_args.rs:91]: this doc is not useful, please remove

File: bin/reth/Cargo.toml [hunk 11]
```diff
@@ -56,6 +56,7 @@ reth-nippy-jar.workspace = true
 
 # crypto
 alloy-rlp.workspace = true
+alloy-chains.workspace = true
```
Comment (mattsse) [inline bin/reth/Cargo.toml:59]: this is not needed here

File: bin/reth/src/cli/db_type.rs [hunk 12]
```diff
@@ -101,7 +101,7 @@ impl DatabaseInstance {
 #[cfg(test)]
 mod tests {
     use super::*;
-    use reth_primitives::Chain;
+    use alloy_chains::Chain;
```
Comment (mattsse) [inline bin/reth/src/cli/db_type.rs:104]: this should still use from pritimitives

File: bin/reth/src/commands/debug_cmd/merkle.rs [hunk 13]
```diff
@@ -278,7 +278,7 @@ impl Command {
                     let clean_result = merkle_stage.execute(&provider_rw, clean_input);
                     assert!(clean_result.is_ok(), "Clean state root calculation failed");
                     if clean_result.unwrap().done {
-                        break;
+                        break
```
Comment (mattsse) [inline bin/reth/src/commands/debug_cmd/merkle.rs:281]: please add back, unrelated changes

File: bin/reth/src/commands/import.rs [hunk 14]
```diff
@@ -215,7 +215,7 @@ mod tests {
     fn parse_common_import_command_chain_args() {
         for chain in SUPPORTED_CHAINS {
             let args: ImportCommand = ImportCommand::parse_from(["reth", "--chain", chain, "."]);
-            assert_eq!(args.chain.chain, chain.parse().unwrap());
+            assert_eq!(&args.chain.chain, &chain.parse::<alloy_chains::Chain>().unwrap());
```
Comment (mattsse) [inline bin/reth/src/commands/import.rs:218]: this shouldn't be required

File: bin/reth/src/commands/node/mod.rs [hunk 15]
```diff
@@ -253,7 +253,7 @@ mod tests {
     fn parse_common_node_command_chain_args() {
         for chain in SUPPORTED_CHAINS {
             let args: NodeCommand = NodeCommand::<()>::parse_from(["reth", "--chain", chain]);
-            assert_eq!(args.chain.chain, chain.parse().unwrap());
+            assert_eq!(args.chain.chain, chain.parse::<alloy_chains::Chain>().unwrap());
```
Comment (mattsse) [inline bin/reth/src/commands/node/mod.rs:256]: same

**Review by mattsse — CHANGES_REQUESTED** [review]
there are lots of unrelated changes (semicolon), please discard those.

and we don't want to add the alloy chains dep everywhere instead we should continue to re-export Chain from the primitives crate

**Comment by 1010adigupta** [comment]
> there are lots of unrelated changes (semicolon), please discard those.
> 
> and we don't want to add the alloy chains dep everywhere instead we should continue to re-export Chain from the primitives crate

File: bin/reth/src/commands/debug_cmd/merkle.rs (hunk 13 above)
Comment (1010adigupta) [reply bin/reth/src/commands/debug_cmd/merkle.rs]: my formatter mistakenly did this, have corrected

File: bin/reth/src/commands/import.rs (hunk 14 above)
Comment (1010adigupta) [reply bin/reth/src/commands/import.rs]: .parse() is unable to infer type here, due which code is not compiling, @Rjected mentioned above to use this solution, if there is any other way you can surely tell me

File: bin/reth/src/cli/db_type.rs (hunk 12 above)
Comment (1010adigupta) [reply bin/reth/src/cli/db_type.rs:104]: ig but the db functions are using alloy_chains::Chain, if I use reth_primitives::Chain it will give a mismatch type error

File: bin/reth/src/commands/node/mod.rs (hunk 15 above)
Comment (1010adigupta) [reply bin/reth/src/commands/node/mod.rs]: same query as above

**Comment by 1010adigupta** [comment]
> there are lots of unrelated changes (semicolon), please discard those.
> 
> and we don't want to add the alloy chains dep everywhere instead we should continue to re-export Chain from the primitives crate

I have removed the semicolon issue, can you give more details on the second one?

File: bin/reth/Cargo.toml (hunk 11 above)
Comment (1010adigupta) [reply bin/reth/Cargo.toml:59]: but I am using alloy-chains in bin/reth, if I remove this, I am unable to use alloy_chains::Chain?
How to proceed here?

File: bin/reth/src/cli/db_type.rs (hunk 12 above)
Comment (mattsse) [reply bin/reth/src/cli/db_type.rs:104]: it should be the same exact type.

File: bin/reth/Cargo.toml (hunk 11 above)
Comment (mattsse) [reply bin/reth/Cargo.toml:59]: these additional deps are not required, because `Chain` is still reexported:

https://github.com/paradigmxyz/reth/blob/main/crates/primitives/src/lib.rs#L54-L54

so none of this should be required.

**Review by mattsse — CHANGES_REQUESTED** [review]
sorry I overlooked that this PR didn't actually remove the all the reth_primitive::Chain types: 

https://github.com/paradigmxyz/reth/blob/main/crates/primitives/src/chain/mod.rs#L99-L99

which was the actual goal of this PR

**Review by mattsse — APPROVED** [review]
great!

I simply removed the reth-primitive chain def and replaced it with the reexport.

now this is complete, ty!

File: bin/reth/src/args/utils.rs [hunk 16]
```diff
@@ -17,7 +17,7 @@ use reth_primitives::{DEV, GOERLI, HOLESKY, MAINNET, SEPOLIA};
 
 #[cfg(feature = "optimism")]
 /// Chains supported by op-reth. First value should be used as the default.
-pub const SUPPORTED_CHAINS: &[&str] = &["base", "base_goerli", "base_sepolia"];
+pub const SUPPORTED_CHAINS: &[&str] = &["base", "base-goerli", "base-sepolia"];
```
Comment (gakonst) [inline bin/reth/src/args/utils.rs:20]: this is a breaking change in the CLI right, we should communicate that in the release notes

Comment (mattsse) [reply bin/reth/src/args/utils.rs:20]: yeah but this now mirrors what op-geth accepts so this is good change
